### ISSUE TITLE
Linux: Option to enable MiddleClickAutoscroll

### ISSUE THREAD

**Issue by pointydev** [Vencord/Vesktop#171]
Due to X using middle click as a keybind to paste from clipboard, Chromium disables middle click scrolling by default. However, this functionality can be re-enabled using `--enable-blink-features=MiddleClickAutoscroll` if a user so desires (as is my case when using [XMousePasteBlock](https://github.com/milaq/XMousePasteBlock)). Would be great to have an in-app option to configure this instead of having to apply it using desktop files or via the command line.

Thanks,
pointy

**Comment by CosmicHeron** [comment]
Maybe an alternative could be how Visual Studio Code reads flags from `.config/code-flags.conf` but instead make it into something like `vencord-flags.conf`?

**Comment by zoobporsor** [comment]
how do i enable autoscroll?

**Comment by TorMechia** [comment]
@zoobporsor just start vesktop with the `--enable-blink-features=MiddleClickAutoscroll` as provided by pointydev.

If launching from terminal, just add it as an argument.

To launch vesktop with autoscroll from desktop, just add the argument to the desktop entry. For KDE just rightclick on the vesktop entry when you search it up > edit application > application > add the argument to the "Arguments" menu entry. Steps should be similar for other desktop environments.

**Comment by Jire** [comment]
> @zoobporsor just start vesktop with the `--enable-blink-features=MiddleClickAutoscroll` as provided by pointydev.
> 
> If launching from terminal, just add it as an argument.
> 
> To launch vesktop with autoscroll from desktop, just add the argument to the desktop entry. For KDE just rightclick on the vesktop entry when you search it up > edit application > application > add the argument to the "Arguments" menu entry. Steps should be similar for other desktop environments.

Even better is making a file in `~/.config/electron-flags.conf` with the flags you want, which all Electron apps will then detect. This is what's in mine:

```
--enable-features=WaylandWindowDecorations
--ozone-platform-hint=auto
--enable-blink-features=MiddleClickAutoscroll
```

But this approach still does not disable pasting with the middle-click button! If anybody has a solution for that (it might potentially be an Electron flag?) to do that would be amazing! (Oh and before you mention, I have it disabled system-wide as much as I can through GNOME Tweaks but I'm using Wayland so I can't use something like a clipboard-clear script nor would I want to do this approach as it could potentially remove important info in my clipboard!)

**Comment by Charlignon** [comment]
A possible solution would be to port [this plugin](https://github.com/hackermare/BD-AutoScroll from BetterDiscord, which would enable autoscroll without the need for the blink feature flag. But this should probably be done at Vencord's level directly.

The only related issue in Vencord is [this one](https://github.com/Vendicated/Vencord/issues/1271) as far as I can tell, but it has been dismissed. I'll try to reopen it with more details and see what happen.

> If anybody has a solution for that (it might potentially be an Electron flag?)
@Jire Unfortunately no the electron/chromium support of this flag is [very minimal](https://issues.chromium.org/issues/40811836) outside of windows. Also this is a bit out of topic, but you can maybe try [this](https://askubuntu.com/a/1079832/1863696) if you haven't already

**Comment by venturiq** [comment]
Installing the PTB version of Vesktop solved the middle click paste issue and enabled autoscroll automatically for me.

**Comment by geopic** [comment]
> Installing the PBT version of Vesktop solved the middle click paste issue and enabled autoscroll automatically for me.

@venturiq What is PBT?

**Comment by venturiq** [comment]
> > Installing the PBT version of Vesktop solved the middle click paste issue and enabled autoscroll automatically for me.
> 
> @venturiq What is PBT?

I think it stands for Public Beta Test. If you right click on the Vesktop icon it should let you click "Reset Vesktop" from there you can change from stable to PBT.

**Comment by Boxersteavee** [comment]
> But this approach still does not disable pasting with the middle-click button! If anybody has a solution for that 
KDE on Wayland now has an option to disable middle-click paste!

Go to System Settings > General Behaviour > Middle Click > De-select "Pastes selected text"

**Comment by velamyGH** [comment]
fwiw, ive found that the kde setting does not work most of the time. id say a plugin would help a ton not only to fix the inconsistent pasting but also enabling autoscroll would be huge since the `--enable-blink-features=MiddleClickAutoscroll` thing doesnt work too consistently for me either.

**Comment by amenokagaseo** [comment]
Electron only respects Wayland compositor middle click drivers if the flag `--ozone-platform=wayland` (or rather, as long as Ozone receives relevant data to switch to Wayland mode) is set, otherwise it runs under XWayland and gives fuck all about what Wayland has to say.

Setting this flag, alongside `--enable-blink-features=MiddleClickAutoscroll` will result in expected behaviour inline with Windows.

> But this approach still does not disable pasting with the middle-click button! If anybody has a solution for that (it might potentially be an Electron flag?) to do that would be amazing! (Oh and before you mention, I have it disabled system-wide as much as I can through GNOME Tweaks but I'm using Wayland so I can't use something like a clipboard-clear script nor would I want to do this approach as it could potentially remove important info in my clipboard!)

@Jire from my testing this works as expected now and middle click paste will no longer pierce through your settings on Wayland applications. YMMV of course.

**Comment by Jire** [comment]
> Electron only respects Wayland compositor middle click drivers if the flag --ozone-platform=wayland (or rather, as long as Ozone receives relevant data to switch to Wayland mode) is set, otherwise it runs under XWayland and gives fuck all about what Wayland has to say.
> 
> Setting this flag, alongside --enable-blink-features=MiddleClickAutoscroll will result in expected behaviour inline with Windows.
> 
> > But this approach still does not disable pasting with the middle-click button! If anybody has a solution for that (it might potentially be an Electron flag?) to do that would be amazing! (Oh and before you mention, I have it disabled system-wide as much as I can through GNOME Tweaks but I'm using Wayland so I can't use something like a clipboard-clear script nor would I want to do this approach as it could potentially remove important info in my clipboard!)
> 
> [@Jire](https://github.com/Jire) from my testing this works as expected now and middle click paste will no longer pierce through your settings on Wayland applications. YMMV of course.

I tested this- it does not work. Middle click still pastes

**Comment by amenokagaseo** [comment]
> I tested this- it does not work. Middle click still pastes

Weird, works on my machine in both Gnome and Plasma sessions. Have you tried `gsettings set org.gnome.desktop.interface gtk-enable-primary-paste false`? The Gnome Tweaks toggle only affects GTK applications if I recall correctly.

**Comment by ynot01** [comment]
I disabled middle click pasting via Linux Mint (22.1, Cinnamon(GNOME)) settings and Firefox uses middle-click scrolling like normal but Vesktop still retains the "paste selection" functionality that I do not desire, even after restarting it

**Comment by Vendicated** [comment]
Afaik, the gnome tweaks setting only works for Wayland apps. If you're running Vesktop through XWayland (default), use [XMousePasteBlock](https://github.com/milaq/XMousePasteBlock). If you're running it through wayland (by running `vesktop --ozone-platform-hint=auto`), gnome tweaks should work

**Comment by ynot01** [comment]
XMousePasteBlock is successful in blocking pastes but unfortunately does not enable scroll functionality

I am also running X11

**Comment by Vendicated** [comment]
because you need to add the aforementioned flag to your vesktop

**Comment by ynot01** [comment]
silly me

**Comment by velamyGH** [comment]
>  the --enable-blink-features=MiddleClickAutoscroll thing doesnt work too consistently for me either.

yeah okay i *think* the inconsistency was from my autostart using the wrong .desktop file or something, for now i think it works with those two flags after making sure i have the correct one in `~/.config/autostart/`. thought it might help someone else here if they did the same idk

**Comment by Vendicated** [comment]
I don't think this is something Vesktop should have a toggle for, as it also requires the user to disable middle click paste in their system

I documented it in [the vesktop wiki](https://vesktop.dev/linux/tips/#enable-middle-click-auto-scroll), that should be good enough

**Comment by Whiplache50** [comment]
> I don't think this is something Vesktop should have a toggle for, as it also requires the user to disable middle click paste in their system
> 
> I documented it in [the vesktop wiki](https://vesktop.dev/linux/tips/#enable-middle-click-auto-scroll), that should be good enough

For me this link leads to a 404 error. However, I can find the page manually on the wiki. But following the instructions listed there doesn't work. When I try to add the --enable-blink-features=MiddleClickAutoscroll flag, I get this error:
`error: Unknown option --enable-blink-features=MiddleClickAutoscroll`

I really think this should be a toggle in vesktop itself, if possible.

**Comment by Dragon-Mom** [comment]
Same as Whiplache here. Neither of these work and the wiki link is a 404. Linux Firefox has this has a toggle so it's a bit unfortunate that it doesn't really seem possible on any version of Discord on Linux, makes it feel a lot clunkier to use than Windows

**Comment by Vendicated** [comment]
the updated wiki link is https://vesktop.dev/wiki/linux/tips/#enable-middle-click-auto-scroll

**Comment by Svnipni** [comment]
Link from vendicated is where it's at.

If you don't want to disable `X11 windowing system` in flatseal, you can try this:

Your `vesktop.desktop` file might look different depending on your distro. The wiki-link refers to GNOME. On my end (PopOS 24 .04 / COSMIC DE) it was `dev.vencord.Vesktop.desktop` in  `.local/share/flatpak/app/dev.vencord.Vesktop/x86_64/stable/{some_numbers}/export/share/applications/`

Just add `--enable-blink-features=MiddleClickAutoscroll` to the end of the `Exec=` path (after `dev.vencord.Vesktop @@u %U @@`) to permanently disable the middle-click paste and enable autoscroll in vesktop.

**Comment by Jire** [comment]
I still could not get middle click paste disabled (ArchLinux with GNOME on Wayland)

What are ALL the flags you guys are using?

@amenokagaseo @Svnipni @Vendicated

**Comment by Svnipni** [comment]
After another update, it seems the .desktop file was reverted. I made the change again along with the Flatseal trick. I can use autoscroll, but it's still registering paste on middleclick on top of it. I'll keep tinkering and will update if I figure something out.

**Comment by Tenrys** [comment]
@Svnipni Did you ever find any solution to this? Last I checked I heard that Chromium fixed this but it seems to not be working or something?

https://issues.chromium.org/issues/399412221#comment19 mentions 144.0.7543.0, but my Vesktop is running on 144.0.7559.134 and it's still going on.
